### ISSUE TITLE
[REVIEW]: Fast k-medoids Clustering in Rust and Python

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#4183]
**Submitting author:** @kno10 (<a href="http://orcid.org/0000-0001-9143-4880">Erich Schubert</a>)
**Repository:** <a href="https://github.com/kno10/python-kmedoids.git" target ="_blank">https://github.com/kno10/python-kmedoids.git</a>
**Version:** 0.2.0
**Editor:** @mikldk
**Reviewer:** @timClicks, @TahiriNadia
**Archive:** Pending

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/4d6eb4cebdad66a96ebf153f986376f3/status.svg)](https://joss.theoj.org/papers/4d6eb4cebdad66a96ebf153f986376f3)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/4d6eb4cebdad66a96ebf153f986376f3"><img src="https://joss.theoj.org/papers/4d6eb4cebdad66a96ebf153f986376f3/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/4d6eb4cebdad66a96ebf153f986376f3/status.svg)](https://joss.theoj.org/papers/4d6eb4cebdad66a96ebf153f986376f3)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/kno10/python-kmedoids.git" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@timClicks & @TahiriNadia, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @mikldk know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

## Review checklist for @timClicks
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/kno10/python-kmedoids.git">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [ ] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [ ] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [ ] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/kno10/python-kmedoids.git">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Contribution and authorship:** Has the submitting author (@kno10) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [ ] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @TahiriNadia
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/kno10/python-kmedoids.git">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [ ] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [ ] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [ ] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/kno10/python-kmedoids.git">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Contribution and authorship:** Has the submitting author (@kno10) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [ ] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @timClicks, @TahiriNadia it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.06 s (186.4 files/s, 18247.1 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                           2             85            206            211
Rust                             1              9             77            174
Markdown                         2             29              0             64
reStructuredText                 1             48             34             57
TOML                             2              4              0             40
YAML                             2              4              1             29
make                             1              1              0              4
-------------------------------------------------------------------------------
SUM:                            11            180            318            579
-------------------------------------------------------------------------------


Statistical information for the repository 'c55cad42f8224e43fd62b6df' was
gathered on 2022/02/19.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Erich Schubert                   8           596             94          100.00

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Erich Schubert              502           84.2          1.4                8.57
```

**Comment by whedon** [comment]
PDF failed to compile for issue #4183 with the following error: 
```
 Can't find any papers to compile :-(

```

**Comment by mikldk** [comment]
@whedon generate pdf from branch JOSS-draft

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch JOSS-draft. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.04183/joss.04183/10.21105.joss.04183.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.04183/joss.04183/10.21105.joss.04183.pdf) :page_facing_up: :point_left:

**Comment by mikldk** [comment]
@whedon check references from branch JOSS-draft

**Comment by whedon** [comment]
```
Attempting to check references... from custom branch JOSS-draft
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1002/9780470316801.ch2 is OK
- 10.1137/0137041 is OK
- 10.1007/978-3-030-32047-8_16 is OK
- 10.1016/j.is.2021.101804 is OK
- 10.21105/joss.01230 is OK
- 10.1287/opre.16.5.955 is OK
- 10.1111/j.1538-4632.1979.tb00674.x is OK
- 10.1007/s10852-005-9022-1 is OK
- 10.1147/sj.22.0129 is OK
- 10.1016/j.eswa.2008.01.039 is OK
- 10.1093/bioinformatics/btp163 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by mikldk** [comment]
@timClicks, @TahiriNadia: Thanks for agreeing to review. 

Please note that both the repos should be reviewed (<https://github.com/openjournals/joss-reviews/issues/4096#issuecomment-1020333611>):

* The engine: <https://github.com/kno10/rust-kmedoids>
* Python interface: <https://github.com/kno10/python-kmedoids>

Please carry out your review in this issue by updating the checklist above and giving feedback in this issue. The reviewer guidelines are available here: <https://joss.readthedocs.io/en/latest/reviewer_guidelines.html>. If possible create issues (and cross-reference) in the submission's repository to avoid too specific discussions in this review thread. 

If you have any questions or concerns please let me know.

@kno10: As you see both repos will be reviewed. This also means that when the repos are ready, you will have to prepare a combined archive (e.g. [zenodo.org](https://zenodo.org/)) of both repositories.

**Comment by mikldk** [comment]
@timClicks, @TahiriNadia, can you please give a brief status of your review? This is not to rush you, merely to give me an impression of the progress and time-frame.

**Comment by mikldk** [comment]
@timClicks, @TahiriNadia, can you please give a brief status of your review? This is not to rush you, merely to give me an impression of the progress and time-frame.

**Comment by mikldk** [comment]
@timClicks, @TahiriNadia, can you please give a brief status of your review? This is not to rush you, merely to give me an impression of the progress and time-frame.

**Comment by timClicks** [comment]
I will try to have my review completed this week. Sorry for the lack of communication!

**Comment by TahiriNadia** [comment]
I will definitely give a brief overview of my review. Sorry too, this is my first time writing a review on JOSS and I don't know how it works.

**Comment by mikldk** [comment]
> I will definitely give a brief overview of my review. Sorry too, this is my first time writing a review on JOSS and I don't know how it works.

Thanks. No worries. Let me know if there is anything I can help you with.

**Comment by TahiriNadia** [comment]
@whedon generate pdf

**Comment by TahiriNadia** [comment]
@editorialbot generate pdf

**Comment by TahiriNadia** [comment]
@editorialbot check references from branch JOSS-draft

**Comment by mikldk** [comment]
@editorialbot set JOSS-draft as branch

**Comment by mikldk** [comment]
@editorialbot check references

**Comment by mikldk** [comment]
@timClicks, @TahiriNadia: Can you please give a brief status of your review? This is not to rush you, merely to give me an impression of the progress and time-frame.

**Comment by kno10** [comment]
Hi, no updates in a month - any progress? @timClicks @TahiriNadia @mikldk

**Comment by timClicks** [comment]
Sorry about the lack of contact from me! I will finish the review within 24h

**Comment by timClicks** [comment]
@editorialbot generate pdf

**Comment by timClicks** [comment]
Re performance claims - the paper only claims to be faster than the upstream Java implementation. I don't think that the paper should be refused because it hasn't been benchmarked against all _k_-medoids/PAM implementations. Accordingly, I have marked the item as satisfied in my review.

**Comment by TahiriNadia** [comment]
I put my comment on this document via PR. My major comment is about the package installation, it gives me a lot of errors. I asked the authors to correct this point by managing a requirements file. However, the idea of the paper is original and very relevant since it compares the best-known classification algorithms (i.e. kmedoids) in different programming languages (i.e. java, python, Rust, C++) and in classical python libraries (i.e. biopython, sklearn_extra, pyclustering). This paper will be of great interest to the scientific community, as the task of classification is ubiquitous in various spheres of society (imaging, marketing, linguistics and others). Speed is often a critical element, especially when performing simulations of large-scale projects. Along with better documentation of how to install, a user guide and finally a requirements file, I highly recommend this article.

**Comment by timClicks** [comment]
**Re "a statement of need"** The authors describe high performance + ease of use as their motivations. This satisfies the requirements within the Documentation section of the review, but unfortunately not the Software Paper section. It requires a paragraph entitled "Statement of Need". I have submitted an [issue in the Python repository](https://github.com/kno10/python-kmedoids/issues/11).

**Re Community Guidelines** I could not find contributor documentation when I looked. I have submitted an issue in the [Python repository](https://github.com/kno10/python-kmedoids/issues/10) along with some example wording.

**Comment by timClicks** [comment]
> My major comment is about the package installation, it gives me a lot of errors.

Installation worked flawlessly for me today.

**Comment by TahiriNadia** [comment]
> > My major comment is about the package installation, it gives me a lot of errors.
> 
> Installation worked flawlessly for me today.

I'm trying it now and I confirm that it's OK for me too.

**Comment by TahiriNadia** [comment]
> I put my comment on this document via PR. My major comment is about the package installation, it gives me a lot of errors. I asked the authors to correct this point by managing a requirements file. However, the idea of the paper is original and very relevant since it compares the best-known classification algorithms (i.e. kmedoids) in different programming languages (i.e. java, python, Rust, C++) and in classical python libraries (i.e. biopython, sklearn_extra, by clustering). This paper will be of great interest to the scientific community, as the task of classification is ubiquitous in various spheres of society (imaging, marketing, linguistics and others). Speed is often a critical element, especially when performing simulations of large-scale projects. Along with better documentation of how to install, a user guide, and finally a requirements file, I highly recommend this article. => FIX by @kno10

**Comment by mikldk** [comment]
@TahiriNadia:  Can you confirm that you have finished the review and recommend that this paper is now published?

@timClicks: Thanks for [specifying the remaining items](https://github.com/openjournals/joss-reviews/issues/4183#issuecomment-1144396584) for @kno10

**Comment by lale1009** [comment]
> 

Hi @timClicks,
we benchmarked the performance against different implementations of kmedoids in different languages:
<img width="589" alt="Screenshot 2022-06-02 134622" src="https://user-images.githubusercontent.com/17047377/171622499-00c2c1df-b9f1-45ba-aa68-0263df051713.png">

**Comment by TahiriNadia** [comment]
> @TahiriNadia: Can you confirm that you have finished the review and recommend that this paper is now published?
> 

@mikldk: Yes, I confirm that everything is finished on my side and I recommend that this paper be published.

**Comment by lale1009** [comment]
@mikldk and @timClicks: We have finished the [remaining items](https://github.com/openjournals/joss-reviews/issues/4183#issuecomment-1144396584)

**Comment by kno10** [comment]
to see the added statement of need:
@editorialbot generate pdf

**Comment by mikldk** [comment]
@kno10 @lale1009: Thanks. I'll wait on @timClicks have the time to check the items.

**Comment by timClicks** [comment]
@editorialbot generate pdf

**Comment by timClicks** [comment]
@mikldk I recommend this paper for publication.

@kno10 @lale1009 Please accept my apologies for the repeated delays.

**Comment by mikldk** [comment]
@timClicks, @TahiriNadia: Thanks for your reviews.

@kno10:

* Please have a final read though of the paper, checking language etc.
* Have a final check of the proofs with `@editorialbot generate pdf`
* In general, be sure that all versions are correct and that the repo is in the state that should be published.
* Please make a tagged release and archive (e.g. with [Zenodo](https://zenodo.org/)) as described [here](<https://joss.readthedocs.io/en/latest/submitting.html#the-review-process>), and report the version number and archive DOI in this thread. Please **verify** that the archive deposit has the correct metadata (**title** and **author list**), or edit these if that is not the case.

**Comment by kno10** [comment]
@mikldk: I have archived a snapshot of the Rust and Python parts version 0.3.3 at Zenodo as:
Schubert, Erich, & Lenssen, Lars. (2022). Fast k-medoids Clustering in Rust and Python (0.3.3). Zenodo. https://doi.org/10.5281/zenodo.6802320
The version number is 0.3.3 (some small bug fixes were integrated, and the package is now automatically built for more architectures to make it easier to install).
I do not have further changes to the paper itself.

**Comment by mikldk** [comment]
@editorialbot set 10.5281/zenodo.6802320 as archive

**Comment by mikldk** [comment]
@editorialbot recommend-accept

**Comment by danielskatz** [comment]
👋 Hi - I'm the AEiC working on finalizing the publication of this submission - I have one minor change in the paper in https://github.com/kno10/python-kmedoids/pull/15 that should be merged, then we can proceed.

**Comment by kno10** [comment]
Thank you. Merged uppercasing Python.

**Comment by danielskatz** [comment]
@editorialbot accept

**Comment by danielskatz** [comment]
Congratulations to @kno10 (Erich Schubert) and co-author!!

And thanks to @timClicks and @TahiriNadia for reviewing, and to @mikldk for editing!
We couldn't do this without you
